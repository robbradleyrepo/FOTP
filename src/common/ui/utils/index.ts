export * from "./composer";
export * from "./css";
export * from "./helpers";
export * from "./provider";
export * from "./styles";
export * from "./types";
