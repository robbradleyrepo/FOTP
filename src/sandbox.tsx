/* eslint-disable simple-import-sort/imports */

// Import for side effects (`configureXYZ(...)`)
import "../pages/_app";

import Sandbox from "@sss/apollo/testing";
import * as ecommerce from "@sss/ecommerce/testing";

import PageProviders from "./ui/providers";
import introspectionResultData from "../generated/config/fragment-types.json";
import { I18N } from "./config";

export const anonymous = new Sandbox({
  Wrapper: PageProviders,
  globalMocks: [
    ecommerce.mockedResponses.checkout.none,
    ecommerce.mockedResponses.collection.core,
  ],
  introspectionResultData,
  locale: I18N.fallbackLocale,
});

export default anonymous;
