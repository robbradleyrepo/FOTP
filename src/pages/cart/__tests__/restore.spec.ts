import { captureException } from "@sentry/nextjs";
import { CheckoutWithVariants, LineItemType } from "@sss/ecommerce/checkout";
import { CHECKOUT_ID, mockedResponses } from "@sss/ecommerce/testing";
import { waitFor } from "@testing-library/react";
import store from "store/dist/store.modern";
import { mocked } from "ts-jest/utils";

import { anonymous } from "../../../sandbox";
import * as toast from "../../../ui/base/toast";
import * as Restore from "../restore";

jest.mock("@sentry/nextjs", () => ({
  captureException: jest.fn(),
}));
jest.mock("uuid", () => ({ v4: jest.fn(() => "mock-id") }));

const mockedCaptureException = mocked(captureException);

const mockToastController = {
  push: jest.fn(),
};

jest.spyOn(toast, "useToastController").mockReturnValue(
  mockToastController as any // eslint-disable-line @typescript-eslint/no-explicit-any
);

describe("checkoutLineItemsToCartLineItems", () => {
  const NODE_ENV = process.env.NODE_ENV;

  afterEach(() => {
    (process.env.NODE_ENV as unknown) = NODE_ENV;
  });

  const product = mockedResponses.product.productFragment;
  const variant = mockedResponses.product.variantFragment;

  const checkoutLineItem = {
    __typename: "RLineItem",
    frequency: {
      __typename: "RFrequency",
      chargeDelayDays: null,
      orderIntervalFrequency: 30,
      orderIntervalUnit: "DAY",
    },
    id: "E9vEeaxExeZQv38FGeT+UAVTr/o=",
    image:
      "https://cdn.shopify.com/s/files/1/0266/2825/9918/products/the-one-x1.png?v=1589206978",
    linePrice: variant.priceV2,
    price: variant.priceV2,
    product: product,
    productId: product.id,
    productType: "",
    properties: [],
    quantity: 1,
    sku: variant.sku,
    title: product.title,
    variant: variant,
    variantId: variant.id,
    variantTitle: variant.title,
    vendor: "Front Of The Pack",
  };

  const checkout = ({
    id: "checkout-id",
    lineItems: [checkoutLineItem],
  } as unknown) as CheckoutWithVariants;

  it("should map ReCharge checkout line items to cart line items", () => {
    expect(
      Restore.checkoutLineItemsToCartLineItems(checkout)
    ).toMatchSnapshot();
  });

  it("should use the `cartLineItemId`, if available", () => {
    const cartLineItemId = "cart-line-item-id";

    const result = Restore.checkoutLineItemsToCartLineItems(({
      ...checkout,
      lineItems: [
        checkoutLineItem,
        {
          ...checkoutLineItem,
          properties: [{ name: "cartLineItemId", value: cartLineItemId }],
        },
      ],
    } as unknown) as CheckoutWithVariants);

    expect(result[0].id).toBe(checkoutLineItem.id);
    expect(result[1].id).toBe(cartLineItemId);
  });

  it("should avoid assigning duplicate line IDs", () => {
    const cartLineItemId = "cart-line-item-id";

    const corruptCheckout = ({
      ...checkout,
      lineItems: [
        {
          ...checkoutLineItem,
          properties: [{ name: "cartLineItemId", value: cartLineItemId }],
        },
        {
          ...checkoutLineItem,
          properties: [{ name: "cartLineItemId", value: cartLineItemId }],
        },
      ],
    } as unknown) as CheckoutWithVariants;

    expect(mockedCaptureException).not.toHaveBeenCalled();

    // Non-production
    expect(() =>
      Restore.checkoutLineItemsToCartLineItems(corruptCheckout)
    ).toThrowErrorMatchingSnapshot();

    // Production
    (process.env.NODE_ENV as unknown) = "production";

    const result = Restore.checkoutLineItemsToCartLineItems(corruptCheckout);

    expect(result[0].id).toBe(cartLineItemId);
    expect(result[1].id).toBe(checkoutLineItem.id);

    expect(mockedCaptureException.mock.calls).toMatchSnapshot();
  });

  it("should handle containers", () => {
    const cartLineItemId = "cart-line-item-id";
    const containerVariantId = "container-variant-id";

    expect(
      Restore.checkoutLineItemsToCartLineItems(({
        ...checkout,
        lineItems: [
          { ...checkoutLineItem, frequency: null },
          {
            ...checkoutLineItem,
            frequency: null,
            product: {
              ...product,
              containers: {
                otp: {
                  base: null,
                  upgrade: {
                    price: { amount: "15.0", currencyCode: "USD" },
                    variantId: containerVariantId,
                  },
                },
                sub: {
                  base: {
                    price: { amount: "0.0", currencyCode: "USD" },
                    variantId: containerVariantId,
                  },
                  upgrade: null,
                },
              },
            },
            properties: [{ name: "cartLineItemId", value: cartLineItemId }],
          },
          {
            properties: [
              { name: "cartLineItemId", value: cartLineItemId },
              { name: "lineItemType", value: LineItemType.CONTAINER },
            ],
            variantId: containerVariantId,
          },
        ],
      } as unknown) as CheckoutWithVariants)
    ).toMatchSnapshot();
  });

  it("should remove reserved properties and return all others", () => {
    const cartLineItemId = "cart-line-item-id";
    const containerVariantId = "container-variant-id";

    expect(
      Restore.checkoutLineItemsToCartLineItems(({
        ...checkout,
        lineItems: [
          {
            ...checkoutLineItem,
            frequency: null,
            product: {
              ...product,
              containers: {
                otp: {
                  base: null,
                  upgrade: {
                    price: { amount: "15.0", currencyCode: "USD" },
                    variantId: containerVariantId,
                  },
                },
                sub: {
                  base: {
                    price: { amount: "0.0", currencyCode: "USD" },
                    variantId: containerVariantId,
                  },
                  upgrade: null,
                },
              },
            },
            properties: [
              { name: "foo", value: "Lorem ipsum dolor sit amet" },
              { name: "cartLineItemId", value: cartLineItemId },
              { name: "bar", value: "Sed condimentum velit sit amet" },
            ],
            variantId: containerVariantId,
          },
        ],
      } as unknown) as CheckoutWithVariants)
    ).toMatchSnapshot();
  });
});

describe("<CartRestore />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    store.clearAll();
  });

  it("should redirect to the cart page if the checkout is successfully restored", async () => {
    const { router } = await anonymous.page({
      NextPage: Restore,
      mocks: [mockedResponses.checkout.restoreSuccess],
      params: {
        token: CHECKOUT_ID,
      },
      pathname: "/cart/restore",
    });

    await waitFor(() => {
      const cart = store.get("cart");

      expect(cart).toMatchSnapshot();

      expect(router.replace).toHaveBeenCalledWith("/cart");
    });
  });

  it("should redirect to the cart page and display a warning if the checkout has been completed", async () => {
    const { router } = await anonymous.page({
      NextPage: Restore,
      mocks: [
        mockedResponses.checkout.restoreFail,
        mockedResponses.checkout.restoreCompletedCoreSuccess,
      ],
      params: {
        token: CHECKOUT_ID,
      },
      pathname: "/cart/restore",
    });

    await waitFor(() => {
      const cart = store.get("cart");

      expect(cart).toMatchSnapshot();

      expect(router.replace).toHaveBeenCalledWith("/cart");
      expect(mockToastController.push).toHaveBeenCalledWith({
        children: "Your cart has already been completed",
        type: toast.ToastType.WARNING,
      });
    });
  });

  it("should redirect to the cart page and display an error if the checkout does not exist", async () => {
    const { router } = await anonymous.page({
      NextPage: Restore,
      mocks: [
        mockedResponses.checkout.restoreFail,
        mockedResponses.checkout.restoreCompletedCoreFail,
      ],
      params: {
        token: CHECKOUT_ID,
      },
      pathname: "/cart/restore",
    });

    await waitFor(() => {
      const cart = store.get("cart");

      expect(cart).toMatchSnapshot();

      expect(router.replace).toHaveBeenCalledWith("/cart");
      expect(mockToastController.push).toHaveBeenCalledWith({
        children:
          "There was a problem restoring your cart: the link you used is invalid or incomplete",
        type: toast.ToastType.ERROR,
      });
    });
  });

  it("should redirect to the cart page is no token is provided", async () => {
    const { router } = await anonymous.page({
      NextPage: Restore,
      pathname: "/cart/restore",
    });

    await waitFor(() => {
      const cart = store.get("cart");

      expect(cart).toMatchSnapshot();

      expect(router.replace).toHaveBeenCalledWith("/cart");
      expect(mockToastController.push).not.toHaveBeenCalled();
    });
  });
});
