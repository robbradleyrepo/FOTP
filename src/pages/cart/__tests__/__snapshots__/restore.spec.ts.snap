// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartRestore /> should redirect to the cart page and display a warning if the checkout has been completed 1`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page and display a warning if the checkout has been completed 2`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page and display an error if the checkout does not exist 1`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page and display an error if the checkout does not exist 2`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page if the checkout is successfully restored 1`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page if the checkout is successfully restored 2`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {
    "key": "val",
  },
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [
    Object {
      "containerPrices": null,
      "containerUpgrade": false,
      "frequency": Object {
        "chargeDelayDays": null,
        "orderIntervalFrequency": 30,
        "orderIntervalUnit": "DAY",
      },
      "handle": "move",
      "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
      "imageUrl": "https://cdn.shopify.com/mock-image/variant-4.jpg",
      "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzE=",
      "properties": Object {},
      "quantity": 1,
      "subscriptionUnitPrice": Object {
        "amount": "29.99",
        "currencyCode": "USD",
      },
      "subtitle": "Default Title",
      "title": "Move",
      "unitPrice": Object {
        "amount": "34.99",
        "currencyCode": "USD",
      },
      "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8x",
    },
  ],
  "rCheckoutId": "1",
  "version": 4,
}
`;

exports[`<CartRestore /> should redirect to the cart page is no token is provided 1`] = `
Object {
  "checkoutId": null,
  "customAttributes": Object {},
  "discountCode": null,
  "id": "mock-id",
  "lineItems": Array [],
  "rCheckoutId": null,
  "version": 4,
}
`;

exports[`checkoutLineItemsToCartLineItems should avoid assigning duplicate line IDs 1`] = `"Duplicate cart line item ID \\"cart-line-item-id\\" in checkout \\"checkout-id\\""`;

exports[`checkoutLineItemsToCartLineItems should avoid assigning duplicate line IDs 2`] = `
Array [
  Array [
    [Error: Duplicate cart line item ID "cart-line-item-id" in checkout "checkout-id"],
    Object {
      "extra": Object {
        "checkout": Object {
          "id": "checkout-id",
          "lineItems": Array [
            Object {
              "__typename": "RLineItem",
              "frequency": Object {
                "__typename": "RFrequency",
                "chargeDelayDays": null,
                "orderIntervalFrequency": 30,
                "orderIntervalUnit": "DAY",
              },
              "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
              "image": "https://cdn.shopify.com/s/files/1/0266/2825/9918/products/the-one-x1.png?v=1589206978",
              "linePrice": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "price": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "product": Object {
                "__typename": "Product",
                "availableForSale": true,
                "bottomline": Object {
                  "__typename": "ReviewBottomline",
                  "averageScore": 4.5,
                  "totalReviews": 2,
                },
                "bundleUnit": null,
                "bundleUnitPlural": null,
                "caloriesPerUnit": null,
                "compareAtPriceV2": null,
                "containers": null,
                "defaultSelectionSku": null,
                "defaultSelectionSubscription": null,
                "defaultVariantOrder": null,
                "description": "Targeted joint and mobility support",
                "featuredImage": Object {
                  "__typename": "ImageConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "id": "1",
                        "src": "https://cdn.shopify.com/mock-image/product-1.jpg",
                      },
                    },
                  ],
                },
                "handle": "move",
                "hasSubscription": null,
                "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
                "images": Object {
                  "__typename": "ImageConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "1",
                        "url": "https://cdn.shopify.com/mock-image/product-1.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "2",
                        "url": "https://cdn.shopify.com/mock-image/product-2.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "3",
                        "url": "https://cdn.shopify.com/mock-image/product-3.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "4",
                        "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                        "width": 1000,
                      },
                    },
                  ],
                },
                "isSubscriptionOnly": null,
                "listingSku": null,
                "listingSubscription": null,
                "listingSubtitle": null,
                "options": Array [
                  Object {
                    "__typename": "ProductOption",
                    "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0T3B0aW9uLzU4NzMxNjM4NjIwOTQ=",
                    "name": "Title",
                    "values": Array [
                      "Default Title",
                    ],
                  },
                ],
                "preorderShippingEstimate": null,
                "preorderType": null,
                "priceRange": Object {
                  "__typename": "ProductPriceRange",
                  "maxVariantPrice": Object {
                    "__typename": "MoneyV2",
                    "amount": "95.98",
                    "currencyCode": "USD",
                  },
                  "minVariantPrice": Object {
                    "__typename": "MoneyV2",
                    "amount": "35.99",
                    "currencyCode": "USD",
                  },
                },
                "productType": "Pet Vitamins & Supplements",
                "seo": Object {
                  "description": "",
                  "title": "",
                },
                "subscriptionFrequencies": null,
                "subscriptionUnit": null,
                "subtitle": Object {
                  "__typename": "Metafield",
                  "id": "metafield-subtitle-move",
                  "value": "Targeted joint and mobility support",
                },
                "title": "Move",
                "totalInventory": 1000,
                "unit": null,
                "variants": Object {
                  "__typename": "ProductVariantConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ProductVariantEdge",
                      "node": Object {
                        "__typename": "ProductVariant",
                        "availableForSale": true,
                        "barcode": "123",
                        "bundleSize": null,
                        "compareAtPriceV2": Object {
                          "__typename": "MoneyV2",
                          "amount": "39.99",
                          "currencyCode": "USD",
                        },
                        "defaultShippingIntervalFrequency": Object {
                          "__typename": "Metafield",
                          "id": "defaultShippingIntervalFrequency-1",
                          "value": "1",
                        },
                        "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
                        "image": Object {
                          "__typename": "Image",
                          "height": 1000,
                          "id": "4",
                          "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                          "width": 1000,
                        },
                        "listingUsp": null,
                        "priceV2": Object {
                          "__typename": "MoneyV2",
                          "amount": "34.99",
                          "currencyCode": "USD",
                        },
                        "quantityAvailable": 10,
                        "selectedOptions": Array [
                          Object {
                            "__typename": "SelectedOption",
                            "name": "Title",
                            "value": "Default Title",
                          },
                        ],
                        "sku": "move",
                        "subscriptionPriceAmount": Object {
                          "__typename": "Metafield",
                          "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                          "value": "29.99",
                        },
                        "title": "Default Title",
                        "units": Object {
                          "__typename": "Metafield",
                          "id": "gid://shopify/Metafield/1",
                          "value": "60",
                        },
                      },
                    },
                  ],
                },
                "vendor": "Front Of The Pack",
              },
              "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
              "productType": "",
              "properties": Array [
                Object {
                  "name": "cartLineItemId",
                  "value": "cart-line-item-id",
                },
              ],
              "quantity": 1,
              "sku": "move",
              "title": "Move",
              "variant": Object {
                "__typename": "ProductVariant",
                "availableForSale": true,
                "barcode": "123",
                "bundleSize": null,
                "compareAtPriceV2": Object {
                  "__typename": "MoneyV2",
                  "amount": "39.99",
                  "currencyCode": "USD",
                },
                "defaultShippingIntervalFrequency": Object {
                  "__typename": "Metafield",
                  "id": "defaultShippingIntervalFrequency-1",
                  "value": "1",
                },
                "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
                "image": Object {
                  "__typename": "Image",
                  "height": 1000,
                  "id": "4",
                  "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                  "width": 1000,
                },
                "listingUsp": null,
                "priceV2": Object {
                  "__typename": "MoneyV2",
                  "amount": "34.99",
                  "currencyCode": "USD",
                },
                "quantityAvailable": 10,
                "selectedOptions": Array [
                  Object {
                    "__typename": "SelectedOption",
                    "name": "Title",
                    "value": "Default Title",
                  },
                ],
                "sku": "move",
                "subscriptionPriceAmount": Object {
                  "__typename": "Metafield",
                  "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                  "value": "29.99",
                },
                "title": "Default Title",
                "units": Object {
                  "__typename": "Metafield",
                  "id": "gid://shopify/Metafield/1",
                  "value": "60",
                },
              },
              "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
              "variantTitle": "Default Title",
              "vendor": "Front Of The Pack",
            },
            Object {
              "__typename": "RLineItem",
              "frequency": Object {
                "__typename": "RFrequency",
                "chargeDelayDays": null,
                "orderIntervalFrequency": 30,
                "orderIntervalUnit": "DAY",
              },
              "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
              "image": "https://cdn.shopify.com/s/files/1/0266/2825/9918/products/the-one-x1.png?v=1589206978",
              "linePrice": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "price": Object {
                "__typename": "MoneyV2",
                "amount": "34.99",
                "currencyCode": "USD",
              },
              "product": Object {
                "__typename": "Product",
                "availableForSale": true,
                "bottomline": Object {
                  "__typename": "ReviewBottomline",
                  "averageScore": 4.5,
                  "totalReviews": 2,
                },
                "bundleUnit": null,
                "bundleUnitPlural": null,
                "caloriesPerUnit": null,
                "compareAtPriceV2": null,
                "containers": null,
                "defaultSelectionSku": null,
                "defaultSelectionSubscription": null,
                "defaultVariantOrder": null,
                "description": "Targeted joint and mobility support",
                "featuredImage": Object {
                  "__typename": "ImageConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "id": "1",
                        "src": "https://cdn.shopify.com/mock-image/product-1.jpg",
                      },
                    },
                  ],
                },
                "handle": "move",
                "hasSubscription": null,
                "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
                "images": Object {
                  "__typename": "ImageConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "1",
                        "url": "https://cdn.shopify.com/mock-image/product-1.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "2",
                        "url": "https://cdn.shopify.com/mock-image/product-2.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "3",
                        "url": "https://cdn.shopify.com/mock-image/product-3.jpg",
                        "width": 1000,
                      },
                    },
                    Object {
                      "__typename": "ImageEdge",
                      "node": Object {
                        "__typename": "Image",
                        "height": 1000,
                        "id": "4",
                        "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                        "width": 1000,
                      },
                    },
                  ],
                },
                "isSubscriptionOnly": null,
                "listingSku": null,
                "listingSubscription": null,
                "listingSubtitle": null,
                "options": Array [
                  Object {
                    "__typename": "ProductOption",
                    "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0T3B0aW9uLzU4NzMxNjM4NjIwOTQ=",
                    "name": "Title",
                    "values": Array [
                      "Default Title",
                    ],
                  },
                ],
                "preorderShippingEstimate": null,
                "preorderType": null,
                "priceRange": Object {
                  "__typename": "ProductPriceRange",
                  "maxVariantPrice": Object {
                    "__typename": "MoneyV2",
                    "amount": "95.98",
                    "currencyCode": "USD",
                  },
                  "minVariantPrice": Object {
                    "__typename": "MoneyV2",
                    "amount": "35.99",
                    "currencyCode": "USD",
                  },
                },
                "productType": "Pet Vitamins & Supplements",
                "seo": Object {
                  "description": "",
                  "title": "",
                },
                "subscriptionFrequencies": null,
                "subscriptionUnit": null,
                "subtitle": Object {
                  "__typename": "Metafield",
                  "id": "metafield-subtitle-move",
                  "value": "Targeted joint and mobility support",
                },
                "title": "Move",
                "totalInventory": 1000,
                "unit": null,
                "variants": Object {
                  "__typename": "ProductVariantConnection",
                  "edges": Array [
                    Object {
                      "__typename": "ProductVariantEdge",
                      "node": Object {
                        "__typename": "ProductVariant",
                        "availableForSale": true,
                        "barcode": "123",
                        "bundleSize": null,
                        "compareAtPriceV2": Object {
                          "__typename": "MoneyV2",
                          "amount": "39.99",
                          "currencyCode": "USD",
                        },
                        "defaultShippingIntervalFrequency": Object {
                          "__typename": "Metafield",
                          "id": "defaultShippingIntervalFrequency-1",
                          "value": "1",
                        },
                        "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
                        "image": Object {
                          "__typename": "Image",
                          "height": 1000,
                          "id": "4",
                          "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                          "width": 1000,
                        },
                        "listingUsp": null,
                        "priceV2": Object {
                          "__typename": "MoneyV2",
                          "amount": "34.99",
                          "currencyCode": "USD",
                        },
                        "quantityAvailable": 10,
                        "selectedOptions": Array [
                          Object {
                            "__typename": "SelectedOption",
                            "name": "Title",
                            "value": "Default Title",
                          },
                        ],
                        "sku": "move",
                        "subscriptionPriceAmount": Object {
                          "__typename": "Metafield",
                          "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                          "value": "29.99",
                        },
                        "title": "Default Title",
                        "units": Object {
                          "__typename": "Metafield",
                          "id": "gid://shopify/Metafield/1",
                          "value": "60",
                        },
                      },
                    },
                  ],
                },
                "vendor": "Front Of The Pack",
              },
              "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
              "productType": "",
              "properties": Array [
                Object {
                  "name": "cartLineItemId",
                  "value": "cart-line-item-id",
                },
              ],
              "quantity": 1,
              "sku": "move",
              "title": "Move",
              "variant": Object {
                "__typename": "ProductVariant",
                "availableForSale": true,
                "barcode": "123",
                "bundleSize": null,
                "compareAtPriceV2": Object {
                  "__typename": "MoneyV2",
                  "amount": "39.99",
                  "currencyCode": "USD",
                },
                "defaultShippingIntervalFrequency": Object {
                  "__typename": "Metafield",
                  "id": "defaultShippingIntervalFrequency-1",
                  "value": "1",
                },
                "id": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
                "image": Object {
                  "__typename": "Image",
                  "height": 1000,
                  "id": "4",
                  "url": "https://cdn.shopify.com/mock-image/variant-4.jpg",
                  "width": 1000,
                },
                "listingUsp": null,
                "priceV2": Object {
                  "__typename": "MoneyV2",
                  "amount": "34.99",
                  "currencyCode": "USD",
                },
                "quantityAvailable": 10,
                "selectedOptions": Array [
                  Object {
                    "__typename": "SelectedOption",
                    "name": "Title",
                    "value": "Default Title",
                  },
                ],
                "sku": "move",
                "subscriptionPriceAmount": Object {
                  "__typename": "Metafield",
                  "id": "Z2lkOi8vc2hvcGlmeS9NZXRhZmllbGQvMTI5ODIzNzg3NTgyMjI=",
                  "value": "29.99",
                },
                "title": "Default Title",
                "units": Object {
                  "__typename": "Metafield",
                  "id": "gid://shopify/Metafield/1",
                  "value": "60",
                },
              },
              "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
              "variantTitle": "Default Title",
              "vendor": "Front Of The Pack",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`checkoutLineItemsToCartLineItems should handle containers 1`] = `
Array [
  Object {
    "containerPrices": null,
    "containerUpgrade": false,
    "frequency": null,
    "handle": "move",
    "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
    "imageUrl": "https://cdn.shopify.com/mock-image/variant-4.jpg",
    "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
    "properties": Object {},
    "quantity": 1,
    "subscriptionUnitPrice": Object {
      "amount": "29.99",
      "currencyCode": "USD",
    },
    "subtitle": "Default Title",
    "title": "Move",
    "unitPrice": Object {
      "amount": "34.99",
      "currencyCode": "USD",
    },
    "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
  },
  Object {
    "containerPrices": Object {
      "otp": Object {
        "base": null,
        "upgrade": Object {
          "amount": "15.0",
          "currencyCode": "USD",
        },
      },
      "sub": Object {
        "base": Object {
          "amount": "0.0",
          "currencyCode": "USD",
        },
        "upgrade": null,
      },
    },
    "containerUpgrade": true,
    "frequency": null,
    "handle": "move",
    "id": "cart-line-item-id",
    "imageUrl": "https://cdn.shopify.com/mock-image/variant-4.jpg",
    "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
    "properties": Object {},
    "quantity": 1,
    "subscriptionUnitPrice": Object {
      "amount": "29.99",
      "currencyCode": "USD",
    },
    "subtitle": "Default Title",
    "title": "Move",
    "unitPrice": Object {
      "amount": "34.99",
      "currencyCode": "USD",
    },
    "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
  },
]
`;

exports[`checkoutLineItemsToCartLineItems should map ReCharge checkout line items to cart line items 1`] = `
Array [
  Object {
    "containerPrices": null,
    "containerUpgrade": false,
    "frequency": Object {
      "chargeDelayDays": null,
      "orderIntervalFrequency": 30,
      "orderIntervalUnit": "DAY",
    },
    "handle": "move",
    "id": "E9vEeaxExeZQv38FGeT+UAVTr/o=",
    "imageUrl": "https://cdn.shopify.com/mock-image/variant-4.jpg",
    "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
    "properties": Object {},
    "quantity": 1,
    "subscriptionUnitPrice": Object {
      "amount": "29.99",
      "currencyCode": "USD",
    },
    "subtitle": "Default Title",
    "title": "Move",
    "unitPrice": Object {
      "amount": "34.99",
      "currencyCode": "USD",
    },
    "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
  },
]
`;

exports[`checkoutLineItemsToCartLineItems should remove reserved properties and return all others 1`] = `
Array [
  Object {
    "containerPrices": Object {
      "otp": Object {
        "base": null,
        "upgrade": Object {
          "amount": "15.0",
          "currencyCode": "USD",
        },
      },
      "sub": Object {
        "base": Object {
          "amount": "0.0",
          "currencyCode": "USD",
        },
        "upgrade": null,
      },
    },
    "containerUpgrade": false,
    "frequency": null,
    "handle": "move",
    "id": "cart-line-item-id",
    "imageUrl": "https://cdn.shopify.com/mock-image/variant-4.jpg",
    "productId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzQ1MzU1Mzc1MzMwMDY=",
    "properties": Object {
      "bar": "Sed condimentum velit sit amet",
      "foo": "Lorem ipsum dolor sit amet",
    },
    "quantity": 1,
    "subscriptionUnitPrice": Object {
      "amount": "29.99",
      "currencyCode": "USD",
    },
    "subtitle": "Default Title",
    "title": "Move",
    "unitPrice": Object {
      "amount": "34.99",
      "currencyCode": "USD",
    },
    "variantId": "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0VmFyaWFudC8zMTc0NjY1MDQwNjk5MA==",
  },
]
`;
